_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,291 @@
+
+These module lifecycles can be used to coordinate module initialization dependencies.
+
+```js
+class TodoList extends Module {  
+  async moduleWillInitialize() {
+    console.log(
+      'TodoList -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'TodoList -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'TodoList -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+}
+
+
+class App extends Module{
+  async moduleWillInitialize() {
+    console.log(
+      'App -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'App -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'App -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+}
+
+const todoList = new TodoList();
+
+const app = App.create({
+  modules: {
+    todoList
+  }
+});
+```
+console.log reulst:
+
+```
+App -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitializeSuccess pendding: true ready: false
+App -> moduleWillInitializeSuccess pendding: true ready: false
+TodoList -> moduleDidInitialize pendding: false ready: true
+App -> moduleDidInitialize pendding: false ready: true
+```
+
+### An ideal architecture
+
+![flow chart](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/flow_chart.png)
+
+In a complex front-end application, this may be a more typical modular architecture design that contains the following sections:
+
+- Lifecycle
+- Store Subscriber
+- Event System
+- State
+- Dependency Modules
+- Domain Models
+
+## Conclusion
+
+When you develop using libraries or framework combinations such as React+Redux/React+MobX/Vue+Vuex and so on, it is hoped that `usm` is a nice option for modularization in your application, which may be the important modular jigsaw puzzle that you lack when building libraries such as React/Vue.
+
+In other words, if you use `usm` for OOP architecture design, your system can not only reduce the boilerplate of different state libraries, especially boilerplate more libraries like Redux, which should be very helpful. Most importantly, `usm` can make the modularization of your OOP architecture simple and intuitive, and even `usm` can make your business code compatible with various state libraries, whether Redux/MobX/Vuex or Angular, and if you're using a UI components library that's just as compatible React/Vue/Angular, then your app will quickly and seamlessly use React/Vue/Angular.
```

Comment:
Modules built with `usm` should be free of boilerplates, especially the type that is introduced by state libraries like redux. More importantly, the object-oriented nature of `usm` makes modules simple and intuitive. `usm` also makes your modules compatible with various state libraries and view layers, allowing you to share your business logic libraries across projects regardless of the frameworks they are using.